torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,795 @@
+            cur_node = node
+            while cur_node.target in non_compute_ops:
+                cur_node = cur_node.args[0]
+                partial_strategy = _gen_partial_strategy(mesh)
+                dp_strategy_map[cur_node] = DataParallelStrategy(
+                    NodeType.GRAD, [partial_strategy]
+                )
+            num_param_grad += 1
+            # remove the tag_grad node from graph
+            node.replace_all_uses_with(node.args[0])
+            train_step_graph.graph.erase_node(node)
+
+            if num_param_grad == num_params:
+                # early break if we have already processed all param_grads
+                break
+
+    train_step_graph.recompile()
+
+    # next we forward propagate to mark all the sharding
+    for node in train_step_graph.graph.nodes:
+        if node.op == "placeholder":
+            if "val" not in node.meta:
+                # NOTE: There're certain cases where the placeholder nodes do
+                # not have real tensor values:
+                # 1. optimizer states can be None sometimes, i.e. SGD with
+                #    no momentum, optimizer states populate `momentum` state
+                #    as None, the full graph we get from `compile` would have
+                #    None as the placeholder value
+                # 2. function args might not only contain params or activations,
+                #    but also contain other non-tensor inputs, i.e. the model
+                #    and optimizer instances baked in as a placeholder, there might
+                #    also be some scalar argument which is not a tensor
+                #
+                # For the above cases, we create a NON_TENSOR stratgy so that we
+                # know it's not a tensor and we don't need to shard it
+                dp_strategy_map[node] = DataParallelStrategy(NodeType.NON_TENSOR, [])
+
+            elif placeholder_idx < num_params:
+                # during compilation there's an assumption that the first num_params
+                # placeholders should be parameters
+                shard_strategy = _gen_shard_strategy(mesh, 0)
+                replica_strategy = _gen_replicate_strategy(mesh)
+                dp_strategy_map[node] = DataParallelStrategy(
+                    NodeType.PARAM, [replica_strategy, shard_strategy]
+                )
+
+            elif placeholder_idx < activation_idx:
+                # optimizer states follow the same strategy as
+                # the corresponding parameters
+                placemenet_strategies: List[PlacementStrategy] = []
+
+                replica_strategy = _gen_replicate_strategy(mesh)
+                placemenet_strategies.append(replica_strategy)
+
+                if node.meta["val"].ndim > 0:
+                    # if the state is a scalar, we don't want to shard it
+                    # i.e. param state step is usually a scalar tensor
+                    shard_strategy = _gen_shard_strategy(mesh, 0)
```

Comment:
Yeah agreed, this is actually the `DEFAULT` DataParallelStyle would do, it would need to actually look into the `param` shape to determine whether the param can be sharded evenly or not, or the param is too small to shard, I will do it as a follow up PR. I think the `build_data_parallel_strategies` would still generate `Replicate()` and `Shard(tensor_dim=0)` as a strategy, to determine which one to really use we can probably do it in `mark_data_parallel_sharding`